sklearn/tree/_tree.pyx

```diff
@@ -1477,29 +1496,23 @@ cdef _cost_complexity_prune(unsigned char[:] leaves_in_subtree, # OUT
             r_node[i] = (
                 weighted_n_node_samples[i] * impurity[i] / total_sum_weights)
 
-        # Push root node, using StackRecord.start as node id
-        rc = stack.push(0, 0, 0, -1, 0, 0, 0)
+        ccp_stack.push({"node_idx": 0, "parent": -1})
```

Comment:
I think this comment has value.
```suggestion
        # Push the root node
        ccp_stack.push({"node_idx": 0, "parent": -1})
```

Reply:
Unrelated to this PR: why is `-1` used there for "parent"? Shouldn't it be `_TREE_UNDEFINED`?

Reply:
Yea, this can be changed to `_TREE_UNDEFINED`, as long as we update the while condition here to also be `!= _TREE_UNDEFINED`

https://github.com/scikit-learn/scikit-learn/blob/b0067e0e7e0ae095592bc3a9a8cb7ba9e200c1be/sklearn/tree/_tree.pyx#L1581

I think it's out of scope for this PR, but a good follow up for maintenance.

Reply:
Opened https://github.com/scikit-learn/scikit-learn/pull/22351 to address this concern. It makes sense to use the constants `_TREE_UNDEFINED`.